pkg/cache/objectarraycache/objectarraycache.go

```diff
@@ -0,0 +1,95 @@
+package objectarraycache
+
+import (
+	"context"
+	"time"
+
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+// RefreshFunction is a function that retrieves an array of objects.
+type RefreshFunction[T any] func(ctx context.Context) ([]*T, error)
+
+// ObjectArrayCache is a cache for an array of objects.
+// The object array in the cache instance are valid for a pre-defined period,
+// and are retrieved from the source when they are not valid anymore.
+type ObjectArrayCache[T any] struct {
+	mutex          sync.RWMutex
+	lastRefreshed  time.Time
+	validityPeriod time.Duration
+	objectCache    []*T
+	refreshFn      RefreshFunction[T]
+}
+
+// NewObjectArrayCache returns a cache for an object array where the cached
+// objects are retrieved using refreshFn and are valid for validityPeriod.
+func NewObjectArrayCache[T any](validityPeriod time.Duration, refreshFn RefreshFunction[T]) *ObjectArrayCache[T] {
+	return &ObjectArrayCache[T]{
+		validityPeriod: validityPeriod,
+		refreshFn:      refreshFn,
+	}
+}
+
+// GetObjects retrieves the array of objects, either from cache (when not
+// expired), or using the refresh function (when the cache is expired).
+func (c *ObjectArrayCache[T]) GetObjects(ctx context.Context) ([]*T, error) {
+	objects, valid := c.getObjectsFromCache()
```

Comment:
I think this solution is suboptimal. We're already in the position where the objects that we're using may be stale. If they are stale past X timeframe, then we should kick off a refresh but we should not block a request and synchronously refresh the backing objects making the request "consistent". We've already accepted that the objects in the cache may not be consistent and thats okay. 

Along the same thought process as above, a 100 requests to getObjectsFromCache and returns !valid for all of them. Then they all get to refreshFn and wait for the query to finish instead of just completing with the current data in the cache

Reply:
Indeed. I'll check how to ensure only one concurrent refresh is triggered once the cache expires.